### ISSUE TITLE
Potential conflicted rules between S602 and S603

### ISSUE THREAD

**Issue by zeshuaro** [charliermarsh/ruff#4045]
After bumping ruff from 0.0.261 to 0.0.262, I'm getting conflicted errors from the rules S602 and S603. See below for the minimal code snippet:

```python
from subprocess import PIPE, Popen

proc = Popen(["/foo/bar"], stdout=PIPE, stderr=PIPE, shell=False)
```

I have been setting `shell=False` as per the S602 rule. However, running this snippet with ruff 0.0.262 errors with the S603 rule, which suggests to set `shell=True`. And if I do that, then the code is failing the S602 rule.

You can find the GitHub Action logs [here](https://github.com/zeshuaro/telegram-pdf-bot/actions/runs/4753738626/jobs/8445683417?pr=1508) with the above error introduced in 0.0.262.

And you can find my Ruff config [here](https://github.com/zeshuaro/telegram-pdf-bot/blob/master/pyproject.toml#L58-L112).

**Comment by dhruvmanila** [comment]
They're here :)

https://github.com/charliermarsh/ruff/blob/cb762f4cad62000e8cd5fa26fec35b228e3bc5d8/crates/ruff/src/rules/flake8_bandit/rules/shell_injection.rs#L252-L253

**Comment by evanrittenhouse** [comment]
Gotcha - I can submit a fix to make these mutually exclusive if that makes sense? @dhruvmanila

**Comment by dhruvmanila** [comment]
I'm confused between these two as they're the exact opposite of each other. If we try to make them mutually exclusive, then which one to detect first?

I think we should actually look into why these two rules exists in the first place and then decide what the fix should be.

Relevant docs:
* B602: https://bandit.readthedocs.io/en/latest/plugins/b602_subprocess_popen_with_shell_equals_true.html
* B603: https://bandit.readthedocs.io/en/latest/plugins/b603_subprocess_without_shell_equals_true.html

---
(A bit late here so calling it a day :))

**Comment by evanrittenhouse** [comment]
Wonder if we could just force either 602, 603, or neither? That way we wouldn't have to worry about which one to pick first. Since they're opposites, I think it makes sense to prevent selecting both

**Comment by zeshuaro** [comment]
I think this Issue on bandit might be related to the behaviour we’re seeing here with ruff?
https://github.com/PyCQA/bandit/issues/333

**Comment by evanrittenhouse** [comment]
It think B603 is a lower priority than B602:
>Because this is a lesser issue than that described in subprocess_popen_with_shell_equals_true a LOW severity warning is reported.

If that's the case, in the case where both are enabled we could defer to checking S602 first - if it's triggered, we could short-circuit the S603 check

**Comment by charliermarsh** [comment]
Huh, interesting. So I guess we're consistent with Bandit here? And the only workaround within Bandit is to add a `# nosec` or similar, to allow low-risk `subprocess` calls?

I suppose we could consider removing `S603` altogether. Or, we could modify `S603` to only flag calls with dynamic arguments, like variables, and let all-string calls be considered "safe", which seems to be what they suggest in that issue.

**Comment by spapanik** [comment]
I think that `S603` makes sense if (even one of) the arguments are dynamic, but for static arguments, it doesn't offer much

**Comment by ofek** [comment]
I vote for removal since even non-literal statics trigger:

```python
[sys.executable, '-m', 'ruff', 'rule', '--all']
```

**Comment by ThiefMaster** [comment]
Yes, this rule seems like garbage right now. It triggers on a simple `subprocess.check_output(['/usr/bin/foo'])` which has zero non-hardcoded input...

**Comment by NeilGirdhar** [comment]
Not sure if this should be a new issue, but I think S602 should also trigger unconditionally for `os.popen` as per [this comment](https://discuss.python.org/t/is-time-to-remove-os-module-spawn-functions/55829/13).

**Comment by dhruvmanila** [comment]
@NeilGirdhar I think that's what `S605` does: https://play.ruff.rs/93f469fb-9725-411c-83c0-2c0d04d0ab75

**Comment by NeilGirdhar** [comment]
Ah, sorry, for some reason my test missed this.  How do you see the lint errors in the playground?  I can't seem to find them.

**Comment by dhruvmanila** [comment]
No worries.

> How do you see the lint errors in the playground? I can't seem to find them.

What do you mean? Do you not see any squiggly lines? You'll need to enable to rules in the settings tab which is on the left sidebar. And, you might also need to enable preview mode if the rule isn't stabilized yet.

**Comment by NeilGirdhar** [comment]
Ah, right, I need to hover, my mistake.  Thanks!
